exemptions: finalize the pipeline, everything should work

OriHoch opened the pull request:
* it updates table `procurement_tenders_exemptions`
* on first run it will insert all data, and then it runs by cronjob every day - which will insert new Ids only

akariv commented:
@OriHoch is this ready or still WIP?

OriHoch commented:
still WIP..

OriHoch commented:
@akariv  - done.. could definately be improved, but it should work

I think the main problem is the update, opened a follow-up issue: #34

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/convert_to_documents_resource.py:
```diff
@@ -0,0 +1,53 @@
+
+
+OUTPUT_TABLE_SCHEMA = {
+    "primaryKey": ["publisher_id", "publication_id", "i"],
```
`publication_id` should suffice.

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/convert_to_documents_resource.py:
```diff
@@ -0,0 +1,53 @@
+    "fields": [
+        {"name": "publisher_id", "type": "integer", "required": True},
+        {"name": "publication_id", "type": "integer", "required": True},
+        {"name": "i", "type": "integer", "required": True},
```
Why do we need `i` here?

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/add_publisher_urls_resource.py:
```diff
@@ -15,11 +17,24 @@
              45: []}
 
 def resource_filter(resource_data, parameters):
+    max_pages = os.environ.get("EXEMPTION_PUBLISHERS_MAX_PAGES", "")
```
why do we need the env vars here? Can't we use parameters?

OriHoch replied:
I am using parameters in the pipeline spec, see below

but, I also allow to override some parameters using environment variables - useful for testing / development

in this case, it can also be used to do the initial sync by setting EXEMPTION_PUBLISHER_MAX_PAGES=-1

akariv replied:
I believe that in tests you should be able to provide parameters as well, no?
Either way, for doing the initial sync, having to set an environment variable is not that ideal (what if the process fails somewhere in the middle? then we need to redo everything from scratch?).
A more robust way would be to check if the data for this document already exists, and if not then do the scraping. This will allow in the future more correct behaviours, such as periodically re-scraping records that are still "in progress" (i.e data might still be changing) and not scraping records that have been concluded.
Checking if a document exists could be done e.g. by hitting the DB and collecting the list of existing publications per publisher id (iirc it's not such a big list).

OriHoch replied:
I was refering to manual testing during development, not automated testing

for example, to test with different number of pages in this case, or in the case of limiting to running for only specific publishers
this allows to quickly modify the parameters from command line with environment variables

I'll fix the update method in #34 - I'll add your comment there

akariv replied:
Okay

akariv replied:
But let's rename those env vars so that people don't confuse them with params

OriHoch replied:
I don't understand why they need to be renamed. what would you rename them to?

akariv replied:
e.g DEBUG_EXEMPTION_PUBLISHER_MAX_PAGES or TEST_EXEMPTION_PUBLISHER_MAX_PAGES so that these are not confused to be part of the official API.

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/parse_exemption_data.py:
```diff
@@ -0,0 +1,108 @@
+TABLE_SCHEMA = {
+    # we take the safe route here and use a combination of values to determine the primary key
+    # I'm not sure how consistent the source data is, and how much we should rely on it for reference keys
+    "primaryKey": ["publisher_id", "publication_id"],
```
`publication_id` should suffice.

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/parse_exemption_data.py:
```diff
@@ -0,0 +1,108 @@
+                "end_date": datetime.strptime(source_data["end_date"], "%d/%m/%Y").date() if source_data["end_date"] else None,
+                "decision": source_data["decision"],
+                "page_title": source_data["page_title"],
+                "documents_json": json.dumps(documents)
```
we can simply call this `documents` and set the type to be `array` - so no need for `json.dumps`

OriHoch replied:
:man_facepalming: 
don't know how I missed this data type..

OriHoch replied:
changed, but it means we now require to use postgresql DB (sqlite won't work for array / object types)
so I guess README will need to be updated as quickstart assumes sqlite DB

akariv replied:
That's an issue! https://github.com/frictionlessdata/datapackage-pipelines/issues/68

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/parse_exemption_data.py:
```diff
@@ -0,0 +1,108 @@
+                "publication_id": int(source_data["publication_id"]),
+                "page_url": exemption["url"],
+                "description": source_data["description"],
+                "supplier_id": int(source_data["supplier_id"]),
```
not an int...

OriHoch replied:
I couldn't find any examples of non-numeric supplier id.. both in the legacy DB and in the data I get from scraping

`select * from exemption where supplier_id is not null and supplier_id !~ '^([0-9]{4,9})$' limit 2`
http://data.obudget.org/queries/1123/source#table

akariv replied:
Supplier if could start with a 0 as a significant digit

akariv replied:
If -> id

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/pipeline-spec.yaml:
```diff
@@ -1,34 +1,69 @@
+        resources:
+          exemptions
+        types:
+          documents_json: null
```
why?

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/pipeline-spec.yaml:
```diff
@@ -1,34 +1,69 @@
       parameters:
-        out-path: /var/datapackages/procurement/tenders/exemptions/publishers/
+        out-path: /var/datapackages/procurement/tenders/exemptions
+    # remove the documents json field before updating SQL
```
Why do we want to do that?

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/pipeline-spec.yaml:
```diff
@@ -1,34 +1,69 @@
-        name: "procurement-tenders-exemptions-publisher-urls"
+        name: "procurement-tenders-exemption-documents"
+    # add the exemption documents resource from the previously saved exemption data
     - run: add_resource
```
let's use `load_resource` instead

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/pipeline-spec.yaml:
```diff
@@ -1,34 +1,69 @@
-publisher-urls:
-  dependencies:
-    - pipeline: ./procurement/tenders/exemptions/publishers
+exemption-documents:
```
After reading the code, I don't think we actually need this pipeline - as all exemption document data is attached as a nested object in the main exemption record.

akariv commented:
Could you also please rebase this branch with master?

OriHoch commented:
@akariv I'm having a hard time with the rebase
I've tried normal methods, but it seems commit history is being deleted / modified which makes it very hard to rebase

maybe I'm doing something wrong, could you write your workflow? or link to a workflow which works (taking into account using forks / feature branches and pull requests etc..)

for example, in this case github says it's fine to merge, but I guess it's not, it's a bit confusing

akariv commented:
@OriHoch I'm guessing the reason is that you're pulling and merging from upstream instead of rebasing on a regular basis.
Rebasing with merge commits is indeed problematic.
In this case I'll simply squash your commits onto the master branch.

OriHoch commented:
fixed an unrelated problem with textract dependency which caused travis build to fail - in latest version of textract they introduced a dependency on swig (command line tool) which is only used for documentation. I fixed by locking the textract dependency to last version which worked. [see this issue for more details](https://github.com/deanmalmgren/textract/issues/159)

akariv commented on budgetkey_data_pipelines/pipelines/procurement/tenders/exemptions/download_exemption_pages_data.py:
```diff
@@ -0,0 +1,85 @@
+        if "db-table" in self.parameters:
+            if not hasattr(self, "all_existing_ids"):
+                self.all_existing_ids = self._get_all_existing_ids(self.parameters["db-table"])
+                self.engine = create_engine(os.environ.get("DPP_DB_ENGINE"))
```
seems to be unnecessary?
